controllers/appwrapper_controller.go

```diff
@@ -105,52 +95,86 @@ func (r *AppWrapperReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		klog.Error(err, "unable to fetch appwrapper")
 	}
 
-	factory := machineinformersv1beta1.NewSharedInformerFactoryWithOptions(machineClient, resyncPeriod()(), machineinformersv1beta1.WithNamespace(""))
-	informer := factory.Machine().V1beta1().MachineSets().Informer()
-	msLister = factory.Machine().V1beta1().MachineSets().Lister()
-	machineLister = factory.Machine().V1beta1().Machines().Lister()
+	// Adds finalizer to the appwrapper if it doesn't exist
+	if !controllerutil.ContainsFinalizer(&appwrapper, finalizerName) {
+		controllerutil.AddFinalizer(&appwrapper, finalizerName)
+		if err := r.Update(ctx, &appwrapper); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
 
-	// todo: Move the getOCMClusterID call out of reconcile loop.
-	// Only reason we are calling it here is that the client is not able to make
-	// calls until it is started, so SetupWithManager is not working.
-	if !useMachineSets && ocmClusterID == "" {
-		getOCMClusterID(r)
+	if !appwrapper.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.finalizeScalingDownMachines(ctx, &appwrapper); err != nil {
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(&appwrapper, finalizerName)
+		if err := r.Update(ctx, &appwrapper); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
 	}
 
-	stopper := make(chan struct{})
-	defer close(stopper)
-	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    onAdd,
-		UpdateFunc: onUpdate,
-		DeleteFunc: onDelete,
-	})
-	go informer.Run(stopper)
-	if !cache.WaitForCacheSync(stopper, informer.HasSynced) {
-		klog.Info("Wait for cache to sync")
+	//TODO Investigate isAWPending()
+	demandPerInstanceType := r.discoverInstanceTypes(&appwrapper)
+	//for userRequestedInstanceType := range demandPerInstanceType {
+	if useMachineSets {
+		if reuse {
+			res, err := r.reconcileReuseMachineSet(ctx, &appwrapper, demandPerInstanceType)
+			if err != nil {
+				klog.Infof("Error reconciling MachineSet: %s", err)
```

Comment:
The error must be returned.

Note that's the AppWrapper that is being reconciled here. Anyway, this can be removed, and let controller managed log the reconcile error.